XiEditor/Trace.swift

```diff
@@ -0,0 +1,96 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// A simple mechanism for logging trace events and outputting a file in
+// Chrome tracing format
+
+import Foundation
+
+var globalTrace = Trace()
+
+class Trace {
+    let mutex = UnfairLock()
+    let BUF_SIZE = 100_000
+    var buf: [TraceEntry]
+    var n_entries = 0
+
+    init() {
+        buf = [TraceEntry](repeating: TraceEntry(), count: BUF_SIZE)
+    }
+
+    func trace(_ name: String, _ cat: TraceCategory, _ ph: TracePhase) {
+        mutex.lock()
+        let i = n_entries % BUF_SIZE
+        buf[i].name = name
+        buf[i].cat = cat
+        buf[i].ph = ph
+        buf[i].abstime = mach_absolute_time()
+        pthread_threadid_np(nil, &buf[i].tid)
+        n_entries += 1
+        mutex.unlock()
+    }
+
+    // TODO: more control over where this gets saved
+    func write() {
+        let pid = getpid()
+        let path = "/tmp/xi-trace-\(pid)"
+        if !FileManager().createFile(atPath: path, contents: nil, attributes: nil) {
```

Comment:
Normally there would be no need to init a new FileManager here; instead you would use the shared `FileManager.default` instance.